crates/chain/src/local_chain.rs

```diff
@@ -0,0 +1,184 @@
+    vec::Vec,
+};
+use bitcoin::BlockHash;
+
+use crate::{BlockId, ChainOracle};
+
+#[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct LocalChain {
+    blocks: BTreeMap<u32, BlockHash>,
+}
+
+// [TODO] We need a cache/snapshot thing for chain oracle.
+// * Minimize calls to remotes.
+// * Can we cache it forever? Should we drop stuff?
+// * Assume anything deeper than (i.e. 10) blocks won't be reorged.
+// * Is this a cache on txs or block? or both?
+// [TODO] Parents of children are confirmed if children are confirmed.
+impl ChainOracle for LocalChain {
+    type Error = Infallible;
+
+    fn get_block_in_best_chain(&self, height: u32) -> Result<Option<BlockHash>, Self::Error> {
+        Ok(self.blocks.get(&height).cloned())
+    }
+}
+
+impl AsRef<BTreeMap<u32, BlockHash>> for LocalChain {
+    fn as_ref(&self) -> &BTreeMap<u32, BlockHash> {
+        &self.blocks
+    }
+}
+
+impl From<LocalChain> for BTreeMap<u32, BlockHash> {
+    fn from(value: LocalChain) -> Self {
+        value.blocks
+    }
+}
+
+impl LocalChain {
+    pub fn tip(&self) -> Option<BlockId> {
+        self.blocks
+            .iter()
+            .last()
+            .map(|(&height, &hash)| BlockId { height, hash })
+    }
+
+    /// This is like the sparsechain's logic, expect we must guarantee that all invalidated heights
+    /// are to be re-filled.
+    pub fn determine_changeset<U>(&self, update: &U) -> Result<ChangeSet, UpdateError>
+    where
+        U: AsRef<BTreeMap<u32, BlockHash>>,
+    {
+        let update = update.as_ref();
+        let update_tip = match update.keys().last().cloned() {
+            Some(tip) => tip,
+            None => return Ok(ChangeSet::default()),
+        };
+
+        // this is the latest height where both the update and local chain has the same block hash
+        let agreement_height = update
+            .iter()
+            .rev()
+            .find(|&(u_height, u_hash)| self.blocks.get(u_height) == Some(u_hash))
+            .map(|(&height, _)| height);
+
+        // the lower bound of the range to invalidate
+        let invalidate_lb = match agreement_height {
+            Some(height) if height == update_tip => u32::MAX,
+            Some(height) => height + 1,
+            None => 0,
+        };
+
+        // the first block's height to invalidate in the local chain
+        let invalidate_from = self.blocks.range(invalidate_lb..).next().map(|(&h, _)| h);
+
+        // the first block of height to invalidate (if any) should be represented in the update
+        if let Some(first_invalid) = invalidate_from {
+            if !update.contains_key(&first_invalid) {
+                return Err(UpdateError::NotConnected(first_invalid));
+            }
+        }
+
+        let invalidated_heights = invalidate_from
+            .into_iter()
+            .flat_map(|from_height| self.blocks.range(from_height..).map(|(h, _)| h));
+
+        // invalidated heights must all exist in the update
```

Comment:
Why? Let's suppose I have the localchain `A -> B -> C -> D`, and the update `B -> C'`. In this case the height of D would be invalidated, but I wouldn't have the same height in the update. Why is this wrong?

Reply:
That's a good point. Yeah, I think it should be okay to roll back the tip. The main fear is we drop blocks in the middle.

Reply:
the rule is the update chain must unambiguously connect to the existing chain. That is it must invalidate the block after the one it connects to or connect to the tip (or invalidate the existing tip).

Reply:
> the rule is the update chain must unambiguously connect to the existing chain. That is it must invalidate the block after the one it connects to or connect to the tip (or invalidate the existing tip).

I don't think this rule alone is sufficient with the new data structures. With this rule alone, there could be disappearing heights in the middle. If this happens, we may have a situation where there are 2 confirmed txs where the 2nd tx is confirmed later than the first, but we will consider the first tx unconfirmed, but the 2nd tx confirmed (due to a missing height in the `ChainOracle` that previously existed).

Reply:
I don't get it. How can their be a missing height? The only way to remove a block at a height is to invalidate it so you be definition don't want it there and don't want to consider any transactions in it to be confirmed.

Reply:
```
Original: A B C D
  Update:   B C'  E 
```

I'm pretty sure this is a valid update, and `D` will be removed. In the new logic I introduced, I wanted to ensure we also introduce `D'` in the update. Otherwise, if we have a tx in `E` that spends a tx in `C'`, it will be very odd.

Reply:
After a call with @LLFourn, we agreed that we will allow "missing bocks" where `ChainOracle::is_block_in_chain` can return `None`, thus meaning "I do not know". What we can do from here, is look at the spends of the tx, and if there is a spend that is confirmed to be in chain, we can assume the tx is also in chain.

Reply:
Implemented.